File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/stores/ThreadListStore.kt
package com.belweave.trifecta.core.stores

import com.belweave.trifecta.core.models.ProjectID
import com.belweave.trifecta.core.models.ProjectShell
import com.belweave.trifecta.core.models.ShellStreamItem
import com.belweave.trifecta.core.models.ThreadShell
import com.belweave.trifecta.core.networking.StreamSubscription
import com.belweave.trifecta.core.networking.T3Client
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.Instant

/**
 * Live, reactive store of all projects + threads in the active workspace.
 * Backed by an `orchestration.subscribeShell` subscription that delivers an
 * initial snapshot followed by upsert/remove deltas.
 *
 * Mirrors the iOS `ThreadListStore` behaviour but exposes flows for Compose.
 */
class ThreadListStore {

    private val mutex = Mutex()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var subscription: StreamSubscription? = null
    private var client: T3Client? = null

    private val _projects = MutableStateFlow<List<ProjectShell>>(emptyList())
    val projects: StateFlow<List<ProjectShell>> = _projects.asStateFlow()

    private val _threads = MutableStateFlow<List<ThreadShell>>(emptyList())
    val threads: StateFlow<List<ThreadShell>> = _threads.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()

    suspend fun start(client: T3Client) {
        mutex.withLock { this.client = client }
        try {
            val sub = client.subscribeShell { item ->
                scope.launch { handle(item) }
            }
            mutex.withLock { subscription = sub }
        } catch (t: Throwable) {
            _lastError.value = t.message
        }
    }

    suspend fun stop() {
        val sub = mutex.withLock {
            val current = subscription
            subscription = null
            current
        }
        sub?.cancel()
    }

    fun reset() {
        _projects.value = emptyList()
        _threads.value = emptyList()
        _lastError.value = null
    }

    private suspend fun handle(item: ShellStreamItem) {
        when (item) {
            is ShellStreamItem.Snapshot -> {
                _projects.value = item.snapshot.projects
                _threads.value = item.snapshot.threads.sortedByDescending { it.sortKey() }
            }
            is ShellStreamItem.ProjectUpserted -> {
                val list = _projects.value.toMutableList()
                val idx = list.indexOfFirst { it.id == item.project.id }
                if (idx >= 0) list[idx] = item.project else list.add(item.project)
                _projects.value = list
            }
            is ShellStreamItem.ProjectRemoved -> {
                _projects.value = _projects.value.filterNot { it.id == item.projectId }
                _threads.value = _threads.value.filterNot { it.projectId == item.projectId }
            }
            is ShellStreamItem.ThreadUpserted -> {
                val list = _threads.value.toMutableList()
                val idx = list.indexOfFirst { it.id == item.thread.id }
                if (idx >= 0) list[idx] = item.thread else list.add(0, item.thread)
                _threads.value = list.sortedByDescending { it.sortKey() }
            }
            is ShellStreamItem.ThreadRemoved -> {
                _threads.value = _threads.value.filterNot { it.id == item.threadId }
            }
        }
    }

    fun threads(in_: ProjectID): List<ThreadShell> =
        _threads.value.filter { it.projectId == in_ && it.archivedAt == null }

    fun project(for_: ThreadShell): ProjectShell? =
        _projects.value.firstOrNull { it.id == for_.projectId }

    fun project(id: ProjectID): ProjectShell? =
        _projects.value.firstOrNull { it.id == id }

    private fun ThreadShell.sortKey(): Instant = latestUserMessageAt ?: updatedAt
}

File: trifecta-ios/Trifecta/Core/Stores/ThreadListStore.swift
import Foundation
import Observation

@Observable
final class ThreadListStore {
    var projects: [ProjectShell] = []
    var threads: [ThreadShell] = []
    var lastError: String?
    private(set) var activeThreads: [ThreadShell] = []
    private(set) var archivedThreads: [ThreadShell] = []
    private(set) var mutationCount: Int = 0

    private var subscription: StreamSubscription?
    private weak var client: T3Client?
    private var projectIndex: [ProjectID: ProjectShell] = [:]
    private var activeThreadsByProject: [ProjectID: [ThreadShell]] = [:]

    func start(client: T3Client) async {
        self.client = client
        subscription = try? await client.subscribeShell { item in
            Task(priority: .utility) { [weak self] in
                let mutation = Self.prepareMutation(item)
                await self?.apply(mutation)
            }
        }
    }

    func stop() async {
        if let sub = subscription { await sub.cancel() }
        subscription = nil
    }

    private enum ShellMutation: Sendable {
        case snapshot(projects: [ProjectShell], threads: [ThreadShell])
        case projectUpserted(ProjectShell)
        case projectRemoved(ProjectID)
        case threadUpserted(ThreadShell)
        case threadRemoved(ThreadID)
    }

    private static func prepareMutation(_ item: ShellStreamItem) -> ShellMutation {
        switch item {
        case .snapshot(let snap):
            return .snapshot(projects: snap.projects, threads: sortByRecency(snap.threads))
        case .projectUpserted(_, let project):
            return .projectUpserted(project)
        case .projectRemoved(_, let id):
            return .projectRemoved(id)
        case .threadUpserted(_, let thread):
            return .threadUpserted(thread)
        case .threadRemoved(_, let id):
            return .threadRemoved(id)
        }
    }

    private static func sortByRecency(_ threads: [ThreadShell]) -> [ThreadShell] {
        threads.sorted { lhs, rhs in
            let l = lhs.latestUserMessageAt ?? lhs.updatedAt
            let r = rhs.latestUserMessageAt ?? rhs.updatedAt
            return l > r
        }
    }

    @MainActor
    private func apply(_ mutation: ShellMutation) {
        switch mutation {
        case .snapshot(let projects, let threads):
            self.projects = projects
            self.threads = threads
            rebuildIndexes()
        case .projectUpserted(let project):
            if let i = projects.firstIndex(where: { $0.id == project.id }) {
                projects[i] = project
            } else {
                projects.append(project)
            }
            rebuildProjectIndex()
        case .projectRemoved(let id):
            projects.removeAll { $0.id == id }
            threads.removeAll { $0.projectId == id }
            rebuildIndexes()
        case .threadUpserted(let thread):
            if let i = threads.firstIndex(where: { $0.id == thread.id }) {
                threads[i] = thread
            } else {
                threads.insert(thread, at: 0)
            }
            threads = Self.sortByRecency(threads)
            rebuildThreadIndexes()
        case .threadRemoved(let id):
            threads.removeAll { $0.id == id }
            rebuildThreadIndexes()
        }
        mutationCount &+= 1
    }

    func threads(in projectId: ProjectID) -> [ThreadShell] {
        activeThreadsByProject[projectId] ?? []
    }

    func project(for thread: ThreadShell) -> ProjectShell? {
        projectIndex[thread.projectId]
    }

    func project(id: ProjectID) -> ProjectShell? {
        projectIndex[id]
    }

    @MainActor
    private func rebuildIndexes() {
        rebuildProjectIndex()
        rebuildThreadIndexes()
    }

    @MainActor
    private func rebuildProjectIndex() {
        projectIndex = Dictionary(uniqueKeysWithValues: projects.map { ($0.id, $0) })
    }

    @MainActor
    private func rebuildThreadIndexes() {
        activeThreads = threads.filter { $0.archivedAt == nil }
        archivedThreads = threads.filter { $0.archivedAt != nil }
        var grouped: [ProjectID: [ThreadShell]] = [:]
        for thread in activeThreads {
            grouped[thread.projectId, default: []].append(thread)
        }
        activeThreadsByProject = grouped
    }
}
